tests/unit/acquisition/function/test_active_learning.py

```diff
@@ -0,0 +1,236 @@
+    builder = ExpectedFeasibility(threshold, alpha, delta)
+    acq_fn = builder.prepare_acquisition_function(model)
+    assert acq_fn._get_tracing_count() == 0  # type: ignore
+
+    query_at = tf.linspace([[-10]], [[10]], 100)
+    expected = bichon_ranjan_criterion(model, threshold, alpha, delta)(query_at)
+    npt.assert_array_almost_equal(acq_fn(query_at), expected)
+    assert acq_fn._get_tracing_count() == 1  # type: ignore
+
+    up_acq_fn = builder.update_acquisition_function(acq_fn, model)
+    assert up_acq_fn == acq_fn
+
+    npt.assert_array_almost_equal(acq_fn(query_at), expected)
+    assert acq_fn._get_tracing_count() == 1  # type: ignore
+
+
+@pytest.mark.parametrize("shape", various_shapes() - {()})
+def test_expected_feasibility_builder_raises_on_non_scalar_threshold(
+    shape: ShapeLike,
+) -> None:
+    threshold, alpha, delta = tf.ones(shape), tf.ones(shape), tf.ones(shape)
+
+    with pytest.raises(TF_DEBUGGING_ERROR_TYPES):
+        ExpectedFeasibility(threshold, 1, 1)
+
+    with pytest.raises(TF_DEBUGGING_ERROR_TYPES):
+        ExpectedFeasibility(1, alpha, 1)
+
+    with pytest.raises(TF_DEBUGGING_ERROR_TYPES):
+        ExpectedFeasibility(1, 1, delta)
+
+
+@pytest.mark.parametrize("alpha", [0.0, -1.0])
+def test_expected_feasibility_builder_raises_on_non_positive_alpha(alpha: float) -> None:
+
+    with pytest.raises(TF_DEBUGGING_ERROR_TYPES):
+        ExpectedFeasibility(1, alpha, 1)
+
+
+@pytest.mark.parametrize("delta", [-1, 0, 1.5, 3])
+def test_expected_feasibility_raises_for_invalid_delta(delta: int) -> None:
+    with pytest.raises(TF_DEBUGGING_ERROR_TYPES):
+        ExpectedFeasibility(1, 1, delta)
+
+
+@pytest.mark.parametrize("delta", [1, 2])
+@pytest.mark.parametrize("at", [tf.constant([[0.0], [1.0]]), tf.constant([[[0.0], [1.0]]])])
+def test_expected_feasibility_raises_for_invalid_batch_size(at: TensorType, delta: int) -> None:
+    ef = bichon_ranjan_criterion(QuadraticMeanAndRBFKernel(), 1, 1, delta)
+
+    with pytest.raises(TF_DEBUGGING_ERROR_TYPES):
+        ef(at)
+
+
+@pytest.mark.parametrize(
+    "threshold, at",
+    [
+        (0.0, tf.constant([[0.0]])),
+        (2.0, tf.constant([[1.0]])),
+        (-0.25, tf.constant([[-0.5]])),
+    ],
+)
+@pytest.mark.parametrize("delta", [1])
+@pytest.mark.parametrize("alpha", [0.1, 1, 2])
+def test_bichon_ranjan_criterion(threshold: float, at: tf.Tensor, alpha: float, delta: int) -> None:
+    model = QuadraticMeanAndRBFKernel()
+    actual = bichon_ranjan_criterion(model, threshold, alpha, delta)(at)
+
+    # approach is to sample based on the model and compute the expectation eq directly
+    mean, variance = model.predict(tf.squeeze(at, -2))
+    stdev = tf.sqrt(variance)
+    normal = tfp.distributions.Normal(tf.cast(0, at.dtype), tf.cast(1, at.dtype))
+    samples = normal.sample(1000000)
+    t = (threshold - mean) / stdev
+    expected = tf.reduce_mean(tf.maximum(0, alpha ** delta - tf.abs(t + samples)))
```

Comment:
should be `tf.abs(t + samples) ** delta` ?

Reply:
also, should be `(alpha * stdev) ** delta`??

Reply:
I have used proposition 4 instead, it shouldn't matter as criterions are derived from the proposition, but I'll check

Reply:
ok, eq 31 instead of the proposition works fine it seems, not sure what is going on, but probably not worth the time to dig into it